ledger/object/sequence.go

```diff
@@ -0,0 +1,400 @@
+	FirstSync() (SequenceRecordPosition, error)
+	// Top returns the greatest position for sequence records.
+	Top(ctx context.Context) (SequenceRecordPosition, error)
+	// TopSync returns position for sequence records at last synced pulse.
+	TopSync(ctx context.Context) (SequenceRecordPosition, error)
+}
+
+// SequenceRecordStorage is a union of interfaces to work with records sequence index.
+type SequenceRecordStorage interface {
+	SequenceRecordAccessor
+	SequenceRecordModifier
+	SequenceRecordCursor
+
+	// UpdatePulse makes "upsert" for top pulse number and sets at old pulse descriptor link to next pulse
+	UpdatePulse(ctx context.Context, pn insolar.PulseNumber) error
+}
+
+// SequenceRecordMemory is an in-memory storage for sequence.
+type SequenceRecordMemory struct{}
+
+// SequenceRecordDB is an db storage impl. It saves sequence to disk and does not allow removal.
+type SequenceRecordDB struct {
+	lock    sync.RWMutex
+	db      store.DB
+	top     SequenceRecordPosition
+	records RecordAccessor
+}
+
+// NewSequenceRecordDB creates new DB storage instance.
+func NewSequenceRecordDB(db store.DB) *SequenceRecordDB {
+	return &SequenceRecordDB{db: db}
+}
+
+// Push saves new record-value in storage and increment topIndex position.
+func (s *SequenceRecordDB) Push(ctx context.Context, pn insolar.PulseNumber, id insolar.ID) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	pd, err := s.exactPulse(pn)
+	if err == nil {
+		pd.topIndex++
+	} else {
+		pd = pulseLabel{curr: pn, next: 0, topIndex: 0}
+	}
+	err = s.updatePulseDescriptor(pn, pd)
+	if err != nil {
+		return errors.Wrap(err, "[SequenceRecordIndex] failed to update pulse descriptor")
+	}
+	pos := SequenceRecordPosition{Index: pd.topIndex, PulseNumber: pn}
+	err = s.set(pos, id)
+	if err != nil {
+		return errors.Wrapf(err, "[SequenceRecordIndex] failed to save record at position: %s", pos)
+	}
+	if pos.GreaterThan(s.top) {
+		s.top = pos
+	}
+	return nil
+}
+
+// ForPosition returns sequence record for provided record position.
+func (s *SequenceRecordDB) ForPosition(ctx context.Context, pos SequenceRecordPosition) (SequenceRecord, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	return s.get(ctx, pos)
+}
+
+// FromPosition returns sequence record slice from provided record position with corresponding limit.
+func (s *SequenceRecordDB) FromPosition(ctx context.Context, pos SequenceRecordPosition, limit uint32) ([]SequenceRecord, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	out := []SequenceRecord{}
+	capacity := uint32(0)
+	for {
+		curr, err := s.exactPulse(pos.PulseNumber)
+		if err != nil {
+			return nil, errors.Wrap(err, "[SequenceRecordIndex] failed to get pulse meta")
+		}
+		for ; pos.Index <= curr.topIndex && capacity < limit; pos.Index++ {
+			rec, err := s.get(ctx, pos)
+			if err != nil {
+				return nil, errors.Wrap(err, "[SequenceRecordIndex] failed to get record")
+			}
+			out = append(out, rec)
+			capacity++
```

Comment:
We don't check the limit here. We can overflow it and the `capacity == limit` check will fail and iterate to the end.